compiler/rustc_typeck/src/coherence/orphan.rs

```diff
@@ -233,13 +233,26 @@ impl ItemLikeVisitor<'v> for OrphanChecker<'tcx> {
                 }
             }
 
-            if let ty::Opaque(def_id, _) = *trait_ref.self_ty().kind() {
-                self.tcx
-                    .sess
-                    .struct_span_err(sp, "cannot implement trait on type alias impl trait")
-                    .span_note(self.tcx.def_span(def_id), "type alias impl trait defined here")
-                    .emit();
-            }
+            // Ensure no opaque types are present in this impl header. See issues #76202 and #86411 for examples,
+            // and #84660 where it would otherwise allow unsoundness.
+            trait_ref.substs.fold_with(&mut BottomUpFolder {
+                tcx: self.tcx,
+                ty_op: |ty| {
+                    if let ty::Opaque(def_id, _) = *ty.kind() {
+                        self.tcx
+                            .sess
+                            .struct_span_err(sp, "cannot implement trait on type alias impl trait")
```

Comment:
I have an idea that could help with this span, not sure if it pans out. Create a new visitor struct. Implement [intravisit::Visitor](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/intravisit/trait.Visitor.html) for it, most notably implement the [Visitor::visit_qpath](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/intravisit/trait.Visitor.html#method.visit_qpath) method and use [qpath_res](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.TypeckResults.html#method.qpath_res) to get the DefId. If it is equal to the opaque type's use that span. If no span is found during visiting, fall back to the one you have right now.

First visit the `self_ty` variable, and if no span is found, visit the `tr` variable.

Reply:
IIUC `qpath_res` doesn't work here since we're not in an `Expr` or `Pat` node, but the qpath should be resolved so we can extract its res anyway. However, its def_id is not the same as the opaque type's here.